doc/developer/design/20231127_decoupled_isolated_coordinator.md

```diff
@@ -0,0 +1,487 @@
+
+We organize the existing milestones ([Use-Case Isolation: Milestones
+(internal)](https://www.notion.so/materialize/Use-case-Isolation-Milestones-f1a37b024ea74b7c9d870778c4349fe3)) into three distinct phases.
+
+### Phase 1: Laying the Foundations
+
+This phase is about developing the components that we will need to decouple
+processing from the single event loop. The current abstractions around the
+catalog and timestamp oracle are not shareable between multiple actors.
+Similarly, table writes cannot be done concurrently by multiple actors.
+
+- [Abstraction for Table Transactions](https://github.com/MaterializeInc/materialize/issues/22173)
+- Differential Catalog State:
+  - [First EPIC](https://github.com/MaterializeInc/materialize/issues/20953)
+  - [Second EPIC](https://github.com/MaterializeInc/materialize/issues/22392)
+- [Distributed TimestampOracle](https://github.com/MaterializeInc/materialize/issues/22029)
+
+We are nearing the end of this phase. All of the above workstreams have
+implementations merged behind feature flags and are being refined before we
+start rolling them out to staging and then production.
+
+### Phase 2: Towards a Vertically Scalable Serving Layer
+
+Once we have the new components developed in Phase 1 we can start incrementally
+pulling responsibilities out of the Coordinator event loop. In line with our
+milestones, we will start with Peeks, move on to table writes, and then finally
+decouple the controllers from the event loop.
+
+- Isolated Peeks: Processing of Peeks will no longer interfere with other
+  responsibilities of the Coordinator, for example processing of controller
+  responses or other (potentially blocking things). Controllers will send Peek
+  Responses back to `pgwire` proactively, bypassing the event loop. And, most
+  importantly other responsibilities will not negatively aspect processing
+  Peeks.
+- Isolated Writes: What we have above for Peeks, but for Writes, basically!
+- Isolated Controller Control Loops: A single cluster (and it's controller),
+  that is, for example, sending back a lot of responses no longer negatively
+  interferes with the rest of the Coordinator responsibilities.
+
+TODO: More exciting details about what the benefits of the individual steps
+will be!
+
+At the end of this phase, we will have an isolated and vertically scalable
+serving layer. The current serving layer can not be scaled up by adding more
+resources to the single process/machine because the single event loop is a
+bottleneck. Once responsibilities are decouple we _can_ scale up vertically.
+
+### Phase 3: A Horizontally Scalable Serving Layer, aka. Full Physical Use-Case Isolation
+
+Once the controllers are decoupled from the Coordinator via the CATALOG pTVC,
+we can start moving them out into their own processes or move them along-side
+the cluster/replica processes. Similarly, we can start moving query processing
+to other processes.
```

Comment:
Is my understanding correct that in this end state we still will have only a single controller per cluster/replica? I.e. we don't need to worry about replicas being able to talk to different controllers at the same time?

Reply:
Yes! But I haven't yet fully worked out _where_ that controller will live. The tricky thing is that there can be multiple replicas, and they each can comprise multiple processes.

Couple options:

- A separate `controllerd` process? That would be easiest to reason about. But cloud might not like that.
- It lives inside one specific replica process. Feels icky!
- It's a distributed thing that lives in all replica processes and internally, somehow, does coordination.

This is one of the last things for Phase III (fully distributed) that I really worry about, because I don't have a good answer yet. 🙈 Good thing we don't need to think about it for Phase II because we still have that one central `environmentd` process...

Reply:
One thing that's maybe helpful to think about is how bad do we need the controller's to be healthy in the Phase III world. If we let the controller run in a replica process, or all replica processes, it will become slow when the replicas are overloaded, and unavailable when the replicas are OOMing (or the cluster has not replicas?).